[WIP] add logabsdet(::UmfpackLU) method

ElOceanografo opened the pull request:
Will fix https://github.com/JuliaLang/LinearAlgebra.jl/issues/788.  This new method is a direct translation of the dense `logabsdet(::LU)` method in LinearAlgebra:https://github.com/JuliaLang/julia/blob/3365b4e903d61b628e2070e25c643cdfcf03c600/stdlib/LinearAlgebra/src/lu.jl#L457

It produces the correct answers, but is relatively slow and needs work on optimization.  The main bottleneck is extracting the factors via `umf_extract`, so directly indexing into the `UmfpackLU` struct should speed it up...though I'm not sure how to do that.

stevengj commented on stdlib/SuiteSparse/src/umfpack.jl:
```diff
@@ -406,6 +406,29 @@ for itype in UmfpackIndexTypes
             complex(mx[], mz[])
         end
+
+        function logabsdet(F::UmfpackLU{Float64, $itype})  # return log(abs(det)) and sign(det)
```
Why only `Float64`?

stevengj commented on stdlib/SuiteSparse/test/umfpack.jl:
```diff
@@ -26,6 +26,7 @@ using LinearAlgebra: Adjoint, Transpose, SingularException
             @test (Diagonal(Rs) * A)[p,q] ≈ L * U
 
             det(lua) ≈ det(Array(A))
+            logabsdet(lua) ≈ logabsdet(Array(A))
```
Doesn't there need to be a `@test` in front of these?   Also, you need to check the sign separately, no?

martinholters committed:
Fix a type-instability in sparse `findmin`/`findmax`

The helper function `_findr` would usually return a `Vector` as first
argument, but would use a `SparseMatrixCSC` in the empty case. Fix by
always using `Vector`.

martinholters committed:
Make sparse `hvcat` inferable

This also requires making sparse `vcat` and `hcat` inferable in the
vararg case which in turn requires a different way to determine the
resulting index type, now implemented similar to `promote_eltype`.

ViralBShah commented:
@ElOceanografo Maybe address the comments and get this merged?

vtjnash committed:
make printing of typealias env shorter

Inspired by #35710

Co-authored-by: Stephan Hilb <stephan@ecshi.net>

ElOceanografo commented:
Sorry, lost track of this! Comments were straightforward, but this is still very slow.  Benchmarking on my computer produces the following timings:
```julia
using LinearAlgebra, SparseArrays, BenchmarkTools
n = 2000
pnz = 0.01
A = sprandn(n, n, pnz); A = A'A + I
Adense = Matrix(A)
F = lu(A)

@btime logabsdet($A) # 549.719 ms (91 allocations: 278.57 MiB)
@btime logabsdet($F) # 79.753 ms (25 allocations: 86.26 MiB)
@btime logabsdet($Adense) # 46.192 ms (3 allocations: 30.53 MiB)
```
Profiling shows most of the time in the first line is spent [here](https://github.com/JuliaLang/julia/blob/3365b4e903d61b628e2070e25c643cdfcf03c600/stdlib/SuiteSparse/src/umfpack.jl#L327), while most of the time in the second line is spent extracting the fields from the LU factorization in [this line](https://github.com/JuliaLang/julia/blob/d8971760436f71e0c07f3a499236bc81fbcec3f8/stdlib/SuiteSparse/src/umfpack.jl#L413).  I know next to nothing about the internals of SuiteSparse, maybe someone who does can chime in?  Using the sparse factorization should in principle be faster than converting to a dense matrix and falling back on the default method, right?

ViralBShah commented:
The problem with testing on random sparse matrices is that the LU factorization gets a lot of fill-in, making it almost dense. In this case, the filled up factors end up having 50% nonzeros. In reality, most sparse problems coming from a real application have more structure and significantly less fill-in.

ViralBShah commented:
`UmfpackLU` is an opaque object, but this ends up spending a bunch of time converting the opaque object, when all you really want are the diagonals. I still think this should be ok to merge. Try some experiments with the identity matrix?

ViralBShah commented:
Probably also rebase to master.

vtjnash committed:
additionally drop trailing env, if unneeded

ElOceanografo commented:
Running the same benchmark with `A = spdiagm(0 => ones(n))` is still slower than the dense version (189.999 μs (81 allocations: 1.11 MiB)).  If you know where the `UmfpackLU` object's layout is documented, I can try to figure out how to pick the diagonal values out directly.  But if you think this is worth merging as-is I can go ahead and rebase it.

ViralBShah commented:
That will have to be in the UmfpackLU documentation or maybe even the Umfpack source code. I had trouble compiling this PR (following the gh command line) instructions and did end up rebasing it to master. Trying to get it working so I can try a few examples myself.

vtjnash committed:
Add detach(::Task) method to monitor for Task failures

This adds a convenience method to be able to essentially wrap an
arbitrary task in a try/catch/log block conveniently.

vtjnash committed:
Add a monitor to some detached Tasks

vtjnash committed:
restrict shorter printing of wheres to gensym names

ElOceanografo commented:
I did some digging in the Umfpack documentation, and it should be possible to extract just the `U` matrix and scaling factors `Rs` by passing null values in for the other arguments (see section 13, top of page 94, in the pdf here: https://fossies.org/linux/SuiteSparse/UMFPACK/Doc/UMFPACK_UserGuide.pdf).  That isn't currently possible through the `SuiteSparse.UMFPACK` wrappers, which is probably worth a separate issue.

vchuravy committed:
[LLVM] patch list for LLVM 12

rssdev10 committed:
docs: clarification about the integer division with 'div' (#40253)

schneiderfelipe committed:
Support cis(A) for matrix A (#40194)

vtjnash committed:
Update CSL (#40287)

* deps: upgrade CSL

This adds the `.so` files, particularly for libatomic that we need on FreeBSD

* build: define uninstall targets even for deps that are not always installed

Most commonly csl, llvm-tools, and clang, but there were other ones in
the wrong lists also.

vtjnash committed:
replace broken replaceuserpath function with existing contractuser one (#40274)

Fixes #39674

vtjnash committed:
ENV: avoid using regex, since these strings are not necessarily valid utf8 (#40273)

quinnj committed:
Fix #40318 by using UInt to convert instead of Int on Ptr

quinnj committed:
Fix rounding cases for fixed-width printing in ryu routines

Fixes #40303. When printing values to fixed widths through the
`Ryu.writefixed` or `Ryu.writeexp` routines, we have a "cleanup" section
after a value has been printed to see if it needs to be rounded given
the input precision and width. The core issue was the terminating
condition: it previously only checked if we were at the start of a
buffer or had encountered the `'-'` character. Via Printf formatting,
however, it also allows specifying the `'+'` and `' '` characters to
preceed a formatted number. Hence, in the OP, the `'1'` character was
getting "rounded" up to the `','` character. The fix here is correctly
checking if the `plus` or `space` options were passed to the routine and
if so, include those in our rounding termination check. The original
issue only reported the "plus" case for the `f` format specifier, but
the same bug affects the `e` format specifier and the "space" option.

quinnj committed:
Fix 32-bit

simeonschaub committed:
fix macro hygiene for `@nospecialize(::T)` (#40293)

Arkoniak committed:
in printstyled, support underline and other flags (#40288)

bicycle1885 committed:
docs: fix typos in AbstractDict (#40325)

dlfivefifty committed:
Fix reduce(vcat,...) type inference (#40277) (#40294)

jishnub committed:
Fix mapreduce docstring (#40214)

HenriDeh committed:
adding default argument value to `reset!(stateful)` (#40158)

I think that most of the time when one calls `reset!(::Stateful)` it will be to simply restart from the begin without changing the iterated collection. So I figured that would avoid `reset!(sf, sf.itr)`. 
I also added the use of `reset!` to the documentation examples because it's not documented anywhere. This way users know of its existence.

srz-zumix committed:
cgmemmgr: check rlimit before using map_size_inc to map more memory (#40136)

charmasaur committed:
Fix `eltype` example in Methods section of manual (#40017)

The code as written doesn't work when passing `eltype(Array{T} where T <: Integer)`, because `Array{T} where T <: Integer` is not a subtype of `AbstractArray{Any}` (but it *is* a subtype of `AbstractArray`).

This was confusing when trying to understand why `Base.eltype(Array{T} where T <: Integer)` returned `Any`, so this re-worded the correct example, to make a little more clear where the fallback is defined.

Removed the two other "correct" examples, and replaced them with another warning about a possible mistake (namely for types without supertypes).

KristofferC committed:
remove mention that non-underscored symbols denote public API (#40145)

adamoppenheimer committed:
Update noteworthy-differences.md (#40326)

They to there

bicycle1885 committed:
make radical operators juxtaposable (#40173)

* make radical operators juxtaposable

* fix test

* update NEWS.md [skip ci]

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

stevengj committed:
document rtol and atol for ≈ in at-test macro (#40061)

* document rtol and atol for ≈ in at-test macro

* Update stdlib/Test/docs/src/index.md

Co-authored-by: Stefan Karpinski <stefan.karpinski@post.harvard.edu>

* Update stdlib/Test/docs/src/index.md

Co-authored-by: Stefan Karpinski <stefan.karpinski@post.harvard.edu>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

jishnub committed:
Add a method to similar with Integers (#39933)

aviatesk committed:
make fuzzy scoring more robust against whitespaces (#39932)

Commit:
Verbalize definition of isapprox in docstring (#39765)

* Verbalize definition of isapprox

* More examples for isapprox

AtsushiSakai committed:
doc: add cross references in subset-related operators (#39750)

Sov-trotter committed:
Add hint in case of missing operator (#40304)

* Add hint for missing operators

* Move hint to it's own error type

* Clear the hint handler after displaying hint

* Don't clear the already empty array in tests

* Don't pop the hint handler

* Rename handler

josePereiro committed:
Add register_error_hint for UndefVarError (#39850)

* add show_error_hints for UndefVarError

Hi, I just include the show_error_hints part from #39089
Further relevant discussion in [discourse](https://discourse.julialang.org/t/help-with-register-error-hint/56140/4)

* Update errorshow.jl

* Update Base.jl

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

navidcy committed:
Adds method to summary() for functions (#39173)

BioTurboNick committed:
docs: Reference functions instead of methods (#38986)

By referencing methods sometimes and not just functions, there was some duplication and confusion. For instance, we had `get(::Any, ::Any, ::Any)` method and also a `get` function, the latter of which duplicated the documentation for the former on the same page.

haampie committed:
Use dot and views in QR code (#38973)

The QR-code is quite ancient, and was written without using other linalg routines because views were expensive then. This lets LinearAlgebra to use dot and norm and such, which are likely optimized for special number types.

krishna-vasudev committed:
doc: Add example for NaN comparison (#38953)

BenjaminGalliot committed:
Add test for Unicode named subpatterns and property mixes of scripts and classes. (#40328)

fp4code committed:
Fixed segfault when adding OffsetArray and UniformScaling (#38544)

bkamins committed:
Clean up of Boolean range indexing (#39993)

This PR references the following comments:

https://github.com/JuliaLang/julia/pull/31829/files#r592887561
https://github.com/JuliaLang/julia/pull/31829/files#r592887003
https://github.com/JuliaLang/julia/pull/39962/files#r592888379

stevengj committed:
correct reverseind example for Unicode (#40156)

simeonschaub committed:
fix inconsistent BroadcastFunction behavior (#40330)

fixes #40309

vtjnash committed:
Revert "Fix some inferability issues in SparseArrays" (#40332)

vtjnash committed:
test: fix bad copy in #40304 (#40334)

vtjnash committed:
remove UndefVarError deprecation from v0.5 (#40333)

Looks like this got lost in deprecation removal?

musm committed:
Fix relpath when path and startpath are in the same drive (#40323)

* Fix relpath when path and startpath are in the same drive

When startpath == ".", assume the startpath is in the same drive.
This subsequently required tweaking the existing logic to then
canonicalize the drive casing in instances where the drive casing
differs.

ViralBShah committed:
Use 0.3.13+4 openblas binaries to fix #40279 (#40343)

* Use 0.3.13+4 openblas binaries to fix #40279
* Add tests and doctest fixes for #40343 (#40346)
* Update checksums
* changed a doctest to have nicer output

Co-authored-by: Pablo San-Jose <lekand@gmail.com>

sethaxen committed:
Fix real matrix sqrt and log for edge cases (#40144)

* Add failing tests

* Improve 2x2 real sqrt for scalar diagonal

* Avoid sylvester if zero is a solution

* Avoid unnecessary sqrt

* Use correct variable names

* Avoid erroring due to NaNs

* Exactly adapt Higham's algorithm, with hypot

* Remove unreachable branch

* Add failing tests

* Handle overflow/underflow

* Invert instead of dividing

* Apply d=0 constraint from standardized real schur form

* Handle underflow

* Avoid underflow/overflow in log diagonal blocks

* Add tests for log underflow/overflow

vtjnash committed:
searchsorted ranges: simplify code, fix bug with Unsigned needle

This avoids computing `length` unnecessarily, which often requires a
division, and is therefore often expensive.

vtjnash committed:
argmin ranges bug: you cannot pass an unsorted array to searchsorted

martinholters committed:
Normalize (simplify) `UnionAll`s when used as type parameter (#36211)

vtjnash committed:
ensure require_one_based_indexing can constant-fold for ranges (#40358)

Refs performance issue noted in #38527

oxinabox committed:
Update Emoji Completions (#39111)

* Update Emoji Completions

* =new should overwrite old

omus committed:
Provide an error message for deserialize_msg

omus committed:
Remove unnecessary multiple returns

vtjnash committed:
errorshow: fix bad test from #40304 (#40376)

This conflicted with a test in core.jl for callable numbers

Commit:
Allow enabling multiple external profilers in build (#38741)

JeffBezanson committed:
fix #40249, reshaping SharedArray on another process (#40286)

carstenbauer committed:
Show methods for LBTConfig and LBTLibraryInfo (#40357)

* show methods for LBTConfig and LBTLibraryInfo

* add interface information to LBTConfig printing

* interface information in show fallback of LBTConfig and LBTLibraryInfo

* LBTConfig(...) instead of LBTConfig() for n>3 libs

staticfloat committed:
[build] Add `symlink_system_library` calls for CSL (#40217)

* [build] Add `symlink_system_library` calls for CSL

We differentiate between a `USE_BINARYBUILDER_CSL=0` configuration
(where we copy the CSL libraries into the Julia build tree) versus a
`USE_SYSTEM_CSL=1` configuration (where we symlink them in a manner
similar to other system libararies)

* Expand `JL_PRIVATE_LIBS-1` for versioned library names as well

Without this, we fail to install things like `libgcc_s.so.1` when
installed into the private libdir by `symlink_system_libraries`.

logankilpatrick committed:
doc: update reducedim.jl to fix a few doc strings (#40351)

Closes #40300, and fixes other docs discrepancies

TommyXR committed:
Add `lock` wrapper for Channels (#39312)

Added lock(f, c::Channel) utility function

stevengj committed:
don't use realpath in Sys.which (#40233)

As [discussed recently on discourse](https://discourse.julialang.org/t/weird-sys-which-function/58070), some programs require a certain executable name to function correctly, and behave badly if you call `realpath` to expand symbolic links (potentially changing the program name).

c.f. discussion in #26559, implemented `Sys.which`

hsm207 committed:
test: simple tests for hash_64_32 and hash_32_32 (#35520)

Tests to help improve coverage

mcognetta committed:
Remove a few redundant diagonal `ldiv!` methods (#38484)

Commit:
Add docs for _ as a variable (#40183)

fredrikekre committed:
Add type asserts to (try)parse(::TimeType, ...) to help inference.

staticfloat committed:
Bump libblastrampoline to v3.0.4 (#40381)

This should fix some armv7l issues, as libblastrampoline had the same
issues in its armv7l trampolines as Julia itself had.  It also fixes
some errors in armv7l interface autodetection.

giordano committed:
[Test] Define `broken`/`skip` keyword argument for `@test` (#39322)

* [Test] Define `broken`/`skip` keyword argument for `@test`

They make it less tedious having tests failing consistently only under specific
conditions.

* [Test] Use new `@test ... broken=cond` in some tests

IanButterworth committed:
Allow setting lower bound on Experimental.@optlevel (#39536)

* allow setting lower bound on Experimental.at-optlevel

* rename arg to min-optlevel

* Apply suggestions from code review

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

* more suggestions from code review

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

IanButterworth committed:
Only print a single `LoadError:` prefix for nested LoadErrors during showerror (#38725)

* fold repeated `LoadError :` prints during showerror

* Only print a single LoadError: prefix

IanButterworth committed:
add option to `compilecache` so that Pkg precompilation can not skip different versions of loaded modules (#40345)

* allow Pkg precomp to not skip loaded modules

* rename to ignore_loaded_modules

* arg alignment

IanButterworth committed:
If module is missing a docstring show readme if available (#39093)

* print available readme if no docstring for module

* tweak search & print

* handle pathless modules

* limit readme print to first 200 lines

* review changes

* print message on single line

* handle modules in sysimage that are missing src files

* tweak formatting

* fix docstring test

Seelengrab committed:
Widen type signature of bytes2hex (#39710)

and make it slightly faster!

Also improves error message on hex2bytes! when passing a non-ASCII string,
And ADD compat notice, fix implementation to be more generic in regards to AbstractString.

ro-ble committed:
doc: hint on how to type the ÷ sign (#38762)

The ÷ sign is the only operator in this chapter which has no dedicated key on the keyboard. With the Unicode input being at the very end of the documentation, newcomers probably do not know how to write the integer division sign. Therefore, I added this short hint on how to type it.

blegat committed:
Replace _iszero by iszerodefined && iszero (#38789)

This PR defines an iszerodefined function that custom types implements in case iszero is well-defined even if zero is not defined.

Take 2 of #36194 after it was reverted in #38764. What was missing in #36194 is that in case iszero is not working for a given type, we should drop the structure and fall back to an Array.

JeffBezanson committed:
fix #40337, no error info from serialized TaskFailedException (#40395)

staticfloat committed:
Bump OpenBLAS_jll to `v0.3.13+5` (#40396)

* Bump OpenBLAS_jll to `v0.3.13+5`

This should fix the illegal instruction errors older processors are
running into.

* Fix doctests

* Update checksums

Co-authored-by: Viral B. Shah <viral@mayin.org>

samuela committed:
Add comments to runtests.jl to explain the `net_on`/loopback situation (#40359)

jmert committed:
fix #33413, crash with `cglobal(dlsym(...))` (#37910)

JeffreySarnoff committed:
doc: clarify example returning a tuple (#40353)

andreasnoack committed:
Avoid AbstractString in GitError struct to make extraction inferred. (#40402)

This significantly reduces the number of invalidated methods when
loading ProgressLogging.

FedericoStra committed:
Fix signed integer overflow in `mapreduce_impl` (#38643)

This PR solves #38627 as regards the code in reducedim.jl, reduce.jl and reinterpretarray.jl.

It does not yet solve all the relevant issues with missing.jl: in the code there for _mapreduce and mapreduce_impl, there are still several instances of the construct:
```
while i <= ilast
    ...
    i += 1
end
```
This overflows when ilast == typemax(Int). More seriously, inside the while loops there is often `@inbounds A[i]`, which is then an out-of-bounds read!

jaakkor2 committed:
Bump MACOSX_VERSION_MIN from 10.9 to 10.10 (#40400)

Fixes #39363 (BinaryBuilder already made this change)

fredrikekre committed:
Fix temp install of BugReporting.jl (#40403)

Fixes #40398

vtjnash committed:
avoid calling dlerror unless absolutely necessary (#40392)

Most platforms implement this sensibly, but POSIX does not require that,
so FreeBSD does not implement this sensibly. But we can mostly avoid the
problems on FreeBSD by avoiding calling it unless we absolutely must.

Fixes #39582

JeffBezanson committed:
remove method definition for `Int` in test/core.jl (#40380)

oscardssmith committed:
faster expm1 (#37440)

Faster and slightly more accurate (and all Julia).

StefanKarpinski committed:
docs: more scope lawyering (re #40238) (#40394)

vtjnash committed:
inference: limit single-level nested `Type` signature (#40379)

Previously `type_more_complex` returns `false` for `Type{Type{...}}`
compared against `Type{...}`. Per comment there, this should return `true`,
but the code was ordered incorrect (it was also missing support for
Vararg, after the recent changes).

Fixes #40336

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

mbauman committed:
Dot-broadcasting for short-circuiting ops .&& and .|| (#39594)

I have long wanted a proper fix for issue #5187. It was the very first Julia issue I filed.
This is a shot at such a fix. This PR:

* Enables parsing for `.&&` and `.||`.  They are parsed into `Expr(:call, :.&&, ...)` expressions at the same precedence as their respective `&&` and `||`:
    ```julia-repl
    julia> Meta.show_sexpr(:(a .&& b))
    (:call, :.&&, :a, :b)
    ```

* Unlike all other dotted operators `.op` (like `.+`), the `op`-alone part (`var"&&"`) is not an exported name from Base. As such, this effectively lowers to `broadcasted((x,y)->x && y, ...)`, but instead of using an anonymous function I've named it `Base.andand` and `Base.oror`:
    ```julia-repl
    julia> Meta.@lower a .&& b
    :($(Expr(:thunk, CodeInfo(
        @ none within `top-level scope'
    1 ─ %1 = Base.broadcasted(Base.andand, a, b)
    │   %2 = Base.materialize(%1)
    └──      return %2
    ))))
    ```

* I've used a named function to enable short-circuiting behavior _within the broadcast kernel itself_. In the case that the second argument is a part of the same fused broadcast kernel, it will only evaluate if required:
    ```julia-repl
    julia> mutable struct F5187; x; end

    julia> (f::F5187)(x) = (f.x += x)

    julia> (iseven.(1:4) .|| (F5187(0)).(ones(4)))
    4-element Vector{Real}:
        1.0
     true
        2.0
     true
    ```

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

blegat committed:
Introduce ArrayContainer in performance tips (#36542)

vtjnash committed:
remove unnecessary `@generated` from literal_pow (#40408)

This was added before the Compiler could constant-prop, now it will just get in the way of general inferability.

arvganesh committed:
Implemented bitwise nand and nor (#40339)

jonas-schulze committed:
Add U+2AEA (⫪) and U+2AEB (⫫) to binary operators (#39403)

quildtide committed:
Improve flow of documentation for non-standard strings (#40057)

DilumAluthgeBot committed:
[automated] Bump the Tar stdlib from ac4d442 to 86825ec (#40421)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

felixrehren committed:
add some assorted subtyping tests (#20627)

MichaelHatherly committed:
[REPL] Implement basic pager using TerminalMenus (#38956)

tlienart committed:
regex: remove error when accessing an unset group (#40391)

closes #31456

alyst committed:
Allow any f for sum/minimum/maximum(f, v::AbstractSparseVector) (#29884)

* sum/minimum/maximum(f, v::AbstractSparseVector)

generalize sum/minimum/maximum(abs/abs2, v::AbstractSparseVector) to
arbitrary f

* add broken sum(f, [])==0 tests for reference

kimikage committed:
doc: fix examples with invalid MIME type (#40427)

KristofferC committed:
fix a typo in TOML error message (#40428)

* fix a typo in TOML error message

ViralBShah committed:
Add LBT checksums and bump versions in all the places. (#40429)

* Add LBT checksums
* remove old checksums
* Change LBT version in various places

logankilpatrick committed:
Update contributing guide with new video links (#40426)

* Update contributing guide with new links

* Update CONTRIBUTING.md

dlfivefifty committed:
Fix eigvals(:::SymTridagonal) with longer off-diagonal vector (#40165)

* Fix eigvals(:::SymTridagonal) with longer off-diagonal vector

* Update tridiag.jl

* Update lapack.jl

* Revert "Update lapack.jl"

This reverts commit 5c4047689c42a6fb4292aa3e6ceff3506e770142.

* Update lapack.jl

sostock committed:
Preserve Hessenberg shape when possible (#40039)

* Preserve Hessenberg shape when possible

* Remove obsolete comments

* Add NEWS

* Support for unitful Hessenberg matrices

* Mark some tests for unitful matrices as broken

eltociear committed:
doc: fix typo in io.jl (#40433)

jz2903 committed:
fix some typos (#40434)

Fix typos in comments

DilumAluthgeBot committed:
[automated] Bump the Pkg stdlib from 74098fe7 to 879c21d9 (#40422)

JeffBezanson committed:
fix #32843, codegen attempting `typeof(NULL)` for some union tags (#40365)

also fixes part of #40065

JeffBezanson committed:
fix #39982, reduce cursor artifacts in REPL after #39538 (#40411)

simeonschaub committed:
fix verbose IRCode printing with long file names (#40437)

Keno committed:
Fix wine bootstrap (#40425)

* Fix wine bootstrap

Fixes #39124
Closes #39022

* fix cross-build, and other build issues

* this should probably be part of spawn, if it is needed

* Revert "this should probably be part of spawn, if it is needed"

This reverts commit eb88a9cf06308a1edb27b15cf1dbb0aff8ae980a.

Seems to be needed

* fixup! fix cross-build, and other build issues

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vchuravy committed:
Update `perf` docs

`-k 1` must be passed for jit injection to work later.

jonas-schulze committed:
Add vim bindings to TerminalMenus (#37940)

* `k` to move up
* `j` to move down
* `<space>` as an alternative to `<enter>`

kshyatt committed:
Allow use of 32 element tuples without dynamic allocation

lkapelevich committed:
define dot between AbstractMatrix and UniformScaling (#40250)

* define dot between AbstractMatrix and UniformScaling

* fix equality check

* fix for complex

* add news entry

oscardssmith committed:
improve fastmath for `exp` and friends (#40223)

* improve fastmath for scalars

* fix typo in exp_impl_fast(::Float64)

* fix whitespace

* Rebase on master

* Update base/special/exp.jl

Co-authored-by: Mustafa M <mus-m@outlook.com>

* Update base/special/exp.jl

Co-authored-by: Mustafa M <mus-m@outlook.com>

* address review

Co-authored-by: Mustafa M <mus-m@outlook.com>

tkf committed:
Always trigger a safepoint in process_events (#40473)

yield is already a potential safepoint, this just ensures it is always one

simeonschaub committed:
further parametrize IRShow (#40459)

Ref https://github.com/JuliaDebug/Cthulhu.jl/issues/149. This allows
Ctulhu to further customize how line annotations are printed. Also moves
a hook to skip printing some statements into `show_ir`, which previously
Cthulhu had to duplicate. Since Cthulhu seems to be the only downstream
consumer of `show_ir` https://juliahub.com/ui/RepoSearch?q=show_ir&r=true,
I don't expect this change to be major. I did also try to keep
compatibility for `show_ir_stmt`, which seems to be more commonly used.

simeonschaub committed:
fix #40478: MethodError hint printed twice (#40479)

https://github.com/JuliaLang/julia/pull/40304 accidentally registered
the error hint every time the error was printed.

vtjnash committed:
fix #39982, remove cursor artifacts in REPL after #40411 (#40452)

aviatesk committed:
improve various inferrabilities, fix some errors (#40463)

KristofferC committed:
prevent conj! into uninitialized memory (#40481)

bicycle1885 committed:
fix %a/%A format for zeros

oxinabox committed:
Relax constrain on `lq` to allwo `AbstractMatrix` (#32120)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

logankilpatrick committed:
Add examples for `@text_str` and `@html_str` (#38254)

Co-authored-by: Max Horn <max@quendi.de>

logankilpatrick committed:
Add docstring example to VecOrMat (#38253)

Co-authored-by: kimikage <kimikage.ceo@gmail.com>

vchuravy committed:
[CLI] error on not finding symbol (#38994)

Also removes some excess exports that are not in julia.h, including
intrinsics and the unused `jl_cpuid` function (only defined on Intel
processors).

ranocha committed:
document how to use `@test_logs` to check for no warnings (#39025)

closes #37307

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

KlausC committed:
Extend `Eigen` to keep additional information from `geevx` (#38483)

Closes #13307
Fixes #38335

bramtayl committed:
no LoadError wrapper in macroexpand errors (#38379)

mcabbott committed:
simplify views of adjoint matrices (#39467)

vtjnash committed:
Revert "simplify views of adjoint matrices (#39467)" (#40504)

This reverts commit 2b4c088ee729228e6014a4973c3e9123044532b4.

There seems to be an issue with #39301 causing now test failure, but this PR was newer and thus easier to revert while reviewing the causes. See discussion in #40504.

mcabbott committed:
restrict vec(::Transpose) shortcut to numbers (#40502)

oscardssmith committed:
float16 cbrt, 50% faster (#39441)

Co-authored-by: Kristoffer Carlsson <kcarlsson89@gmail.com>

eltociear committed:
[CLI] Fix typo in loader_lib.c (#40506)

initalize -> initialize

musm committed:
sign executables within installer

sostock committed:
Relax type restriction on `factors` field of `LQPackedQ` (#39665)

dbrueg committed:
Add references to Performance Tips and Workflow Tips to Getting Started. (#40464)

I suspect that some people go to julialang.org, click 'Learn', click
'getting started guide', start reading, and never get to Performance Tips
(or Workflow Tips) because they come quite late in the manual.

(An alternative would be to link to the introduction which already has a link to Performance Tips.
However, I like that the Getting Started page quickly gets to the REPL,
so I don't want to change that.)

musm committed:
Update Julia year in exe

dlfivefifty committed:
Overload permutedims for Diagonal (#39447)

Optimization for performance

abraunst committed:
add spzeros(::Tuple{<:Integer}) (#40515)

dkarrasch committed:
Fix `kron` with `Diagonal` (#40509)

ettersi committed:
Move CHOLMOD.common_struct from Vector{UInt8} to an actual struct (#38919)

* Move common_struct from Vector{UInt8} to an actual struct

* Compare full version numbers

Co-authored-by: Simon Etter <ettersi@nus.edu.sg>

vtjnash committed:
[SparseArrays] similar on sparse matrix returned uninitialized space (#40444)

Co-authored-by: Klaus Crusius <klaus.crusius@web.de>

mcabbott committed:
faster reductions of Transpose, Adjoint, PermutedDimsArray (#39513)

oscardssmith committed:
faster Float16 sinh (#39432)

vtjnash committed:
safepoints are required in any lock than may be used with allocations (#40487)

(which is pretty much all locks)

quildtide committed:
Set DelimitedFiles.readdlm use_mmap default to false for all OSes (#40415)

Increased resilience to edge cases where OS is reported Unix-like but Filesystem is abnormal

Fixes #8891
Fixes #40352

PaulSoderlind committed:
add  `selectdim(A, 2, 3:4)` example (#40456)

This adds a `selectdim(A, 2, 3:4)` example to illustrate that several indices can be used (as this is not entirely clear in the text). Based on a suggestion in https://discourse.julialang.org/t/pick-rows-from-array-ndims-not-known-in-advance/59033/6

aramirezreyes committed:
Added docstring for -> (#40470)

Co-authored-by: Fredrik Ekre <ekrefredrik@gmail.com>

regier21 committed:
Add docs for RegexMatch and keys(::RegexMatch) (#40486)

murari0 committed:
Add `get(::Function, ::ImmutableDict, key)` to resolve #40413 (#40471)

kshyatt committed:
LibGit2: branch missing test in last_error (#39033)

dbrueg committed:
Add scal!(a,X) (#40446)

vtjnash committed:
Revert "LibGit2: branch missing test in last_error (#39033)" (#40530)

This reverts commit a4fc56a57cf04b5aa460e271761d12294598a2fe.

mcabbott committed:
Add many "see also" links to docstrings (#38606)

vtjnash committed:
enable new Float16 methods for sinh and cbrt (#40534)

From #39432 and #39441, these were still using their old definition due
to method overwriting.

dkarrasch committed:
Fix eigvecs(:::SymTridiagonal) with longer off-diagonal vector (#40526)

Co-authored-by: Marcelo Forets <mforets@gmail.com>

sostock committed:
Promote eltype when dividing UniformScaling by matrix (#40528)

vtjnash committed:
docs: mention privacy of type fields in style-guide (#40533)

close #12064
close #7561

Co-authored-by: Spencer Russell <spencer.f.russell@gmail.com>

fredrikekre committed:
SparseArrays: add Printf to test deps. (#40520)

simeonschaub committed:
remove `===` for singleton equality tip (#40517)

With #38905 merged, this shouldn't apply in general anymore. I am sure with `==` being a generic function, there will still be cases where using `===` will lead to more precise inference, but I don't think this is worth specifically mentioning in the performance tips.

mbauman committed:
doc: `take!(::IOBuffer)` may make copies (#27442)

Simple doc change to reflect the implementation.

jch8ri committed:
add some docs for Dates formatters

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

andyferris committed:
Document the guarantees that should be provided by `isequal` (#34798)

tkf committed:
Document what public API means (#35715)

Provides some guidelines, in addition to those added recently by #40533 to the style-guide, per #7561.

jaideep-seth committed:
doc: add `@` to help (#33824)

Co-authored-by: jaideep-seth <jaideepnseth@gmail.com>

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

johnnychen94 committed:
unexport cat_shape (#40549)

DebadityaPal committed:
Transpose and Adjoint support for exp, log and sqrt functions (#39373)

ranocha committed:
at_test_nowarn does not suppress the content of stderr anymore (#39028)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

ArunS-tack committed:
added docstring for single-quotes pair in basedocs.jl (#39609)

One of the operators in #33666 that wasn't previously searchable

blegat committed:
Fix rational division by zero (#40551)

* Fix rational division by zero
* Replace `BigFloat` by `BigInt`

Co-authored-by: Sebastian Stock <42280794+sostock@users.noreply.github.com>

vtjnash committed:
logging: redirect closed streams to stderr/stdout (#40423)

This has the additional benefit of making the initial logger respect
changes to redirect_stderr/stdout, until the user explicitly sets
another stream as the logging destination.

Fix #26798
Fix #38482
Replaces #26920, which provided the idea and most of the implementation

Co-authored-by: Joe Petviashvili <joe@gnom.us>

jmkuhn committed:
Document splitext() with multiple extensions (#40554)

jamesonquinn committed:
docs: add FAQ on parametric invariance (#21261)

bobcassels committed:
docs: provide more detail on some noteworthy differences with Common Lisp (#36685)

Roger-luo committed:
allow creating an undef Diagonal (#38282)

nalimilan committed:
Add docstring for keys(::AbstractArray) (#36073)

This is helpful for users, and it matters because some Base functions may
rely on these guarantees for their correct behavior (e.g. `hash`).

Co-authored-by: Matt Bauman <mbauman@juliacomputing.com>

SimonDanisch committed:
fetch is a noop when on the same worker (#19044)

I was quite confused by this example, since I didn't realize that there is a `getindex(f::Future, args...) = getindex(fetch(f), args...)`.
Because of this I misinterpreted the example (my first conclusion: `remote_call` must magically unpack the `Future`).
Also, it didn't really make clear, that fetch turns into a noop if applied to a `Future` from the same worker.
I tried to add this information in a concise way.

vtjnash committed:
process: make SIGPIPE an always failure (#39574)

This error code does not exist on Windows, so any code that relied on
this was broken (would fail this test) on that platform already. In most
cases, it was actually a serious error too, such as `cp -v` or `tar -v`,
which will indeed fail midway and give corrupt results if you do not read
their output to completion!

Reverts #1469, in essence

vtjnash committed:
remove some exports from Broadcast (#40543)

It seems like these were unintentional (and undocumented). Refs #26919.

dkarrasch committed:
Doc: Mention that dropdims'ed array shares memory (#40544)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
Co-authored-by: Michael Abbott <32575566+mcabbott@users.noreply.github.com>

rikhuijzer committed:
Improve type stability for tryparse VersionNumber (#40557)

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

JeffBezanson committed:
change some `::AbstractString` fields to `::String` (#40569)

simonbyrne committed:
Document assignment destructuring (#30579)

* Document assignment destructuring
* Add reference to slurping assignment (#37410)
* mention swapping variables and other kinds of LHS

Co-authored-by: Jameson Nash <vtjnash@gmail.com>
Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

staticfloat committed:
Fix `llvm-tools` when installing with `LLVM_ASSERTS=1` (#40447)

When installing with `LLVM_ASSERTS=1`, we need our `llvm-tools` to be an
assert build as well.

belamenso committed:
Fix #31368: joinpath works on collections of paths (#38263)

* Fix #31368: joinpath works on collections of paths

* Update base/path.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

* Apply suggestions from code review

* Fixup

Co-authored-by: Mustafa M <mus-m@outlook.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

iagobaapellaniz committed:
make \cdot, \interpunct, and \cdotp equivalent (#25157)

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

Co-authored-by: Steven G. Johnson <stevenj@alum.mit.edu>

staticfloat committed:
[cli] Provide `.type` and `.size` annotations for symbols (#40575)

On ELF platforms, `ld` can print out warnings that certain symbols do
not have a type or size annotation.  This adds the annotations to the
`DEBUGINFO` and `EXPORT` helper macros, which are where this information
was put for the windows assembler.  With this patch, we should eliminate
`ld` warnings such as:

```
warning: type and size of dynamic symbol `jl_symbol' are not defined
```

oscardssmith committed:
faster Float32 and Float16 pow (#40236)

Approximately .5 ULP, relatively fast.
Update float^integer as well

bicycle1885 committed:
avoid excessive inlining of Ryu (#40593)

sostock committed:
Add CHOLMOD error handler (#40589)

* Add CHOLMOD error handler

* Remove some now-obsolete at-isok

* Remove at-isok macro

vtjnash committed:
Random: cut randstring allocation size in half (#40584)

JeffreySarnoff committed:
add `inttype`, `floattype` like `uinttype` (#36526)

abraunst committed:
stricter buffer sizes in SparseMatrixCSC (#40523)

Make length(A.nzval)==nnz(A) and add strict buffer checking (#30662)

* Add sizehint!(::SparseMatrixCSC, args...),
* Fix illegal SparseMatrixCSC construction in cholmod and linalg.
* Remove tests targetting now illegal buffers
* Fix invalid buffer creation in kron and more
* use widelength in sizehint! to cope with large matrices in 32 bit systems

stev47 committed:
implement faster floating-point `isless` (#39090)

* implement faster floating-point `isless`

Previously `isless` relied on the C intrinsic `fpislt` in
`src/runtime_intrinsics.c`, while the new implementation in Julia
arguably generates better code, namely:

 1. The NaN-check compiles to a single instruction + branch amenable
    for branch prediction in arguably most usecases (i.e. comparing
    non-NaN floats), thus speeding up execution.
 2. The compiler now often manages to remove NaN-computation if the
    embedding code has already proven the arguments to be non-NaN.
 3. The actual operation compares both arguments as sign-magnitude
    integers instead of case analysis based on the sign of one
    argument. This symmetric treatment may generate vectorized
    instructions for the sign-magnitude conversion depending on how the
    arguments are layed out.

The actual behaviour of `isless` did not change and apart from the
Julia-specific NaN-handling (which may be up for debate) the resulting
total order corresponds to the IEEE-754 specified `totalOrder`.

While the new implementation no longer generates fully branchless code I
did not manage to construct a usecase where this was detrimental: the
saved work seems to outweight the potential cost of a branch
misprediction in all of my tests with various NaN-polluted data. Also
auto-vectorization was not effective on the previous `fpislt` either.

Quick benchmarks (AMD A10-7860K) on `sort`, avoiding the specialized
algorithm:

```julia
a = rand(1000);
@btime sort($a, lt=(a,b)->isless(a,b));
    # before: 56.030 μs (1 allocation: 7.94 KiB)
    #  after: 40.853 μs (1 allocation: 7.94 KiB)
a = rand(1000000);
@btime sort($a, lt=(a,b)->isless(a,b));
    # before: 159.499 ms (2 allocations: 7.63 MiB)
    #  after: 120.536 ms (2 allocations: 7.63 MiB)
a = [rand((rand(), NaN)) for _ in 1:1000000];
@btime sort($a, lt=(a,b)->isless(a,b));
    # before: 111.925 ms (2 allocations: 7.63 MiB)
    #  after:  77.669 ms (2 allocations: 7.63 MiB)
```


* Remove old intrinsic fpslt code

Co-authored-by: Mustafa Mohamad <mus-m@outlook.com>

jishnub committed:
kron for Diagonal uses kron for components (#40596)

dkarrasch committed:
Make length type match index type in sparse vectors (#39645)

ViralBShah committed:
Create README.md (#40603)

[ci skip]
[skip ci]

rfourquet committed:
add docstring for Base.load_path() (#40582)

Commit:
REPL: Fix incorrect docstring for .. (#40563) (#40607)

This PR fixes #40563 .
This is essentially a mirror of #39975 . Please refer to it for more information.

simeonschaub committed:
Revert "faster Float32 and Float16 pow (#40236)" (#40610)

This reverts commit 1474566ffc1dfe16f643b446f667178b78b56aa6.

KristofferC committed:
remove debug logging for the TOML cache in loading (#40581)

aviatesk committed:
improve inferrabilities by telling the compiler relational invariants (#40594)

Our compiler doesn't understand these relations automatically yet.

rfourquet committed:
show(::MersenneTwister) : use module qualification (#40583)

show(::MersenneTwister) : use module qualification

Instead of always printing `MersenneTwister(...)`, we also prepend
`Random.` when not is scope, e.g.
```julia
julia> import Random; Random.MersenneTwister(0)
Random.MersenneTwister(0)
```

simeonschaub committed:
deprecate `getproperty(::Pairs, s)` (#39448)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

simeonschaub committed:
test package completion if already imported (#36552)

0xAlexS committed:
doc: git checkout 1.6.1 in README (#40633)

rfourquet committed:
add *(::Missing, ::Char) (#40615)

staticfloat committed:
Add some documentation to cli/loaders (#40636)

KristofferC committed:
add some precompile for Artifacts and add a type annotation (#40631)

aviatesk committed:
minor refactor for `Base._which` (#40617)

Two minor changes:
1. look up matching methods in a passed world
2. pass the current world by default

xref: https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/478

IanButterworth committed:
Enable REPL to offer to install missing packages if install hooks are provided (#39026)

ettersi committed:
Make cholmod preserve user-specified permutation (#40560)

* Make cholmod preserve user-specified permutation

* Ensure all changes to CHOLMOD.common are temporary

* Fix spurious comment

Co-authored-by: Simon Etter <ettersi@nus.edu.sg>

staticfloat committed:
Update CSL to eliminate dependence on `clock_gettime()` (#40499)

* Update SuiteSparse checksums to include `aarch64-apple-darwin`

* Update CSL to eliminate dependence on newer macOS support functions

Our compiler support libraries had been incorrectly built to use symbols
available only in macOS 10.12+; this update correctly configures them to
not use `clock_gettime()` and `mkostemp`.

aviatesk committed:
REPL: fix #27184, ensure macro existence before lowering (#40621)

jishnub committed:
Fix showindices for IdentityUnitRange (#40608)

bicycle1885 committed:
add docstring for VecElement (#40019)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

tkf committed:
Include all exported functions in Test.jl documentation (#37361)

eltociear committed:
Fix typo in opaque_closure.c (#40647)

paramters -> parameters

vchuravy committed:
[LLVM] Drop support for versions prior to LLVM 11 (#40634)

waldyrious committed:
mpfr.jl: don't add extra zero when pretty-printing BigFloats (#40513)

* mpfr.jl: don't add extra zero when pretty-printing BigFloats

* Cleanup MPFR tests

- Make blank lines consistent: none around testset blocks,
  but present around standalone tests
- Clarify name of testset: 'precision' --> 'setprecision'
- Merge two redundant testsets on comparison operators

jishnub committed:
Fix reduced_index for Integer UnitRanges having eltypes other than Int (#39999)

* reduced_index for BigInt UnitRanges

* Add tests for Integer subtypes

* use OneTo constructor instead of typeof

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

dbrueg committed:
Make opnorm( 1x0 matrix ) zero (#40483)

* Make opnorm( 1x0 matrix ) zero

Fixes #40370.

* more opnorm tests

* remove convert on svdvals

Co-authored-by: Daniel Karrasch <Daniel.Karrasch@gmx.de>

MilesCranmer committed:
Make edit(f) return methods if multiple signatures available (#39481)

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

sanjibansg committed:
more efficient array indexing with AbstractUnitRange (#39896)

Co-authored-by: Sebastian Stock <42280794+sostock@users.noreply.github.com>

vtjnash committed:
doc: mentions of addenv seemed lacking (#40539)

Co-authored-by: akaysh <ksh.shrm1@gmail.com>
Co-authored-by: Sacha Verweij <sacha.verweij@alumni.stanford.edu>

Co-authored-by: akaysh <ksh.shrm1@gmail.com>
Co-authored-by: Sacha Verweij <sacha.verweij@alumni.stanford.edu>

DilumAluthge committed:
Add tests to make sure that `Base.project_names`, `Base.manifest_names`, `Base.preferences_names`, and `Artifacts.artifact_names` always have the same length (#40618)

* Add tests to make sure that `Base.project_names`, `Base.manifest_names`, `Base.preferences_names`, and `Artifacts.artifact_names` always have the same length

* Add a comment explaining why we have these tests

JeffBezanson committed:
a bit of cleanup to init/loading (#40637)

KristofferC committed:
make sure sortperm on Vector{Float64} is precompiled (#40645)

rfourquet committed:
add "see also" for `strip`, `lstrip` and `rstrip` docstrings (#40648)

jishnub committed:
indexing arrays with abstractunitranges retains indices (#40660)

alhirzel committed:
fix typo in "Rejecting cache file" message (#40662)

tuedo2 committed:
Added test for promotion of same-sized unsigned and signed integers (#39739)

sairus7 committed:
`@lock` export from Base, closes #36441 (#39588)

Commit:
[doc] minor change of filename in getting-started.md (#40674)

JackDunnNZ committed:
Add compat to docstring for single-arg `isapprox` (#40659)

Co-authored-by: Simeon Schaub <simeondavidschaub99@gmail.com>

DilumAluthgeBot committed:
[automated] Bump the Pkg stdlib from 879c21d9 to ad3d6396 (#40684)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

bicycle1885 committed:
show sign for NaNs if required

Commit:
Optimisation of getproperty for UmfpackLU object (avoid copying irrelevant information) (#40663)

jishnub committed:
ndims for AbstractArray{<:Any,N} (#40682)

IanButterworth committed:
update all checksums and pcre rules to generate them (#40688)

* update all checksums and pcre rules to generate them

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

DilumAluthgeBot committed:
[automated] Bump the Pkg stdlib from ad3d6396 to 1a47f445 (#40687)

Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

ElOceanografo committed:
only extract needed fields from UmfpackLU

using new efficient method from #40663

ElOceanografo committed:
Merge branch 'sparse_logabsdet' of https://github.com/ElOceanografo/julia into sparse_logabsdet

ElOceanografo committed:
Merge branch 'master' of https://github.com/JuliaLang/julia into sparse_logabsdet

ViralBShah commented:
Something has gone wrong with this PR.

ElOceanografo commented:
I tried to rebase it to master last night and may have messed it up.  I can try to fix it, or if it's easier I could close it and open a new one with the same changes...

ViralBShah commented:
If its easier to do a new PR, just go for it.

ElOceanografo commented:
Ok, closing this, will open a new one in a second.
